src/Features/Core/Portable/MakeFieldReadonly/AbstractMakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,194 @@
+                    {
+                        nonReadonlyFieldMembers.Add(symbol);
+                    }
+                }
+            }
+
+            var syntaxFactsService = GetSyntaxFactsService();
+            foreach (var syntaxReference in typeSymbol.DeclaringSyntaxReferences)
+            {
+                var typeNode = syntaxReference.SyntaxTree.GetRoot(context.CancellationToken).FindNode(syntaxReference.Span);
+
+                var semanticModelForTree = context.SemanticModel.Compilation.GetSemanticModel(syntaxReference.SyntaxTree);
+                RemoveAssignedSymbols(semanticModelForTree, syntaxFactsService, typeNode, nonReadonlyFieldMembers, context.CancellationToken);
+            }
+
+            foreach (var symbol in nonReadonlyFieldMembers)
+            {
+                var diagnostic = Diagnostic.Create(
+                    GetDescriptorWithSeverity(option.Notification.Value),
+                    symbol.Locations[0]);
+                context.ReportDiagnostic(diagnostic);
+            }
+
+            nonReadonlyFieldMembers.Free();
+        }
+
+        private void RemoveAssignedSymbols(SemanticModel model, ISyntaxFactsService syntaxFactsService, SyntaxNode node, PooledHashSet<IFieldSymbol> unassignedSymbols, CancellationToken cancellationToken)
+        {
+            foreach (var descendant in node.DescendantNodes())
+            {
+                if (unassignedSymbols.Count == 0)
+                {
+                    return;
+                }
+
+                if (!(descendant is TIdentifierNameSyntax name))
+                {
+                    continue;
+                }
+
+                var symbol = model.GetSymbolInfo(descendant).Symbol as IFieldSymbol;
+                if (symbol == null || !unassignedSymbols.Contains(symbol))
+                {
+                    continue;
+                }
+
+                if (!IsMemberOfThisInstance(descendant))
+                {
+                    unassignedSymbols.Remove(symbol);
+                }
+
+                if (IsDescendentOf<TConstructorDeclarationSyntax>(descendant, out var ctorNode))
+                {
+                    var isInAnonymousOrLocalFunction = false;
+                    for (var current = descendant.Parent; current != ctorNode; current = current.Parent)
+                    {
+                        if (syntaxFactsService.IsAnonymousOrLocalFunction(current))
+                        {
+                            isInAnonymousOrLocalFunction = true;
+                            break;
+                        }
+                    }
+
+                    if (isInAnonymousOrLocalFunction)
+                    {
+                        unassignedSymbols.Remove(symbol);
+                    }
+                    else
+                    {
+                        var ctorSymbol = model.GetDeclaredSymbol(ctorNode);
+                        if (!ctorSymbol.ContainingType.Equals(symbol.ContainingType))
```

Comment:
ah, you check the containing type here.  ok :)

Reply:
➡️ No action required.

Reply:
i'd still like a test.  it's an easy case to miss, and it would be helpful in case any of this code gets refactored :)